staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -991,12 +991,14 @@ func (s *store) prepareKey(key string) (string, error) {
 
 // decode decodes value of bytes into object. It will also set the object resource version to rev.
 // On success, objPtr would be set to the object.
-func decode(codec runtime.Codec, versioner storage.Versioner, value []byte, objPtr runtime.Object, rev int64) error {
+func decode(codec runtime.Codec, resource string, key string, versioner storage.Versioner, value []byte, objPtr runtime.Object, rev int64) error {
```

Comment:
document the resource and key parameters are only intended to aid metrics/logging if the decode fails, they must not affect decode behavior at all

Reply:
Sorry I'm not sure I understand it correctly.
Do you mean we should not add parameters to decode() function? If yes, how should we pass in `resource`/`key` to record metric and logging?

Reply:
passing them in is fine, but the parameters should be named and/or documented to make it clear their only intended purpose is for logging

Reply:
I don't like passing these in here, it's extremely confusing, why not make the call sites call a separate function to handle the error?

Reply:
Thank you @liggitt . Clear to me now.

@lavalamp Do you mean keep func `decode()` unchanged, and for each time we call decode(...), wrap it like
```
err = decode(...)
if err != nil {
    recordDecodeError(resource, preparedKey)
}

func recordDecodeError(resource string, key string) {
    metrics.RecordDecodeError(resource)
    klog.V(4).Infof("Decoding %s \"%s\" failed", resource, key)
}
```

Reply:
Yes, like that. Note for listed items, that preparedKey is not going to be helpful at all for debugging.